### DISCUSSION TITLE
Mix HTTP server and client

### DISCUSSION THREAD

**Discussion by jbbcst** [cesanta/mongoose#1620] [Ask questions, get answers] [answered]
Hi all,

I am using the example of http server from Mongoose examples in a test project. Is it possible from this example to communicate with another server using an http client request? Can I use both at the same time in the main?

Thanks in advance

**Comment by scaprile** [comment, marked as answer]
As long as you assign the proper function handler you can have several different connection types. In fact, the [reverse proxy example](https://github.com/cesanta/mongoose/tree/master/examples/http-reverse-proxy) does what you want. Tutorial [here](https://mongoose.ws/tutorials/http-reverse-proxy/)

**Comment by jbbcst** [comment]
Thank you for your response, I looked at the reverse proxy example. Can I use `mg_connect(c->mgr, s_backend_url, fn2, c);` out of the fn function with the same `struct mg_mgr mgr` that it passed to `mg_http_listen(&mgr, s_listen_url, fn, NULL);` ?

**Reply by scaprile** [reply]
If you followed the tutorial, and the other two tutorials it would link you to, you've seen usually there is only one event manager. 
`c->mgr` is initialized to point to `mgr` when the connection structure is allocated (when you call `mg_connect()`). All callbacks in the example are called from the very same event manager when it runs in `main()`: `mg_mgr_poll(&mgr, ...);`

**Comment by jbbcst** [comment]
Ok thank you, I am going to do some tests

**Comment by jbbcst** [comment]
Hello,
I tried something like this:
```
int main(void) {
    struct mg_mgr mgr;
    mg_mgr_init(&mgr);
    mg_http_listen(&mgr, "http://192.168.1.33:8000", fn, NULL);  // Create listener
    for (;;)
    {
        mg_mgr_poll(&mgr, 1000);                         // Event loop
        if (sendToClient) myFn(mgr.conns, NULL);
    }
    mg_mgr_free(&mgr);                                        // Cleanup
    return 0;
}
```

And myFn:

```
void myFn(struct mg_connection* c, void* fn_data)
{
    struct mg_connection* c2 = (mg_connection*)fn_data;
    sendToClient++;
    // Client request, create backend connection Note that we're passing
    // client connection `c` as fn_data for the created backend connection.
    c2 = mg_connect(c->mgr, "192.168.0.28:8000", fn2, c);
    if (c2 == NULL) {
        mg_error(c, "Cannot create backend connection");
    }
    else {
        if (mg_url_is_ssl(s_backend_url.c_str())) {
            struct mg_tls_opts opts = { .ca = "ca.pem" };
            mg_tls_init(c2, &opts);
        }
        c->fn_data = c2;
        //forward_request(hm, c2);
        c2->is_hexdumping = 1;
        struct mg_str host = mg_url_host("192.168.1.16:8000");
        int content_length = s_post_data ? strlen(s_post_data) : 0;
        mg_printf(c,
            "%s %s HTTP/1.0\r\n"
            "Host: %.*s\r\n"
            "Content-Type: octet-stream\r\n"
            "Content-Length: %d\r\n"
            "\r\n",
            s_post_data ? "POST" : "GET", mg_url_uri("192.168.1.16:8000"), (int)host.len,
            host.ptr, content_length);
        mg_send(c, s_post_data, content_length);        
    }
}
```
The mg_connect() seems to work properly because I see a connection on the server side. 
However the http request Get is not sent, is the right way to join a server?
Do you have an example where I can "correctly" mix a server that behaves as a client?

**Reply by scaprile** [reply]
If you need separate server and client, just follow the [server](https://mongoose.ws/tutorials/http-server/) and [client](https://mongoose.ws/tutorials/http-client/) tutorials (they are referenced in the reverse proxy tutorial I linked in my first response) and join them. A server and a client are independent, each one has its event handler `fn`, you just call the client connect function when you need to (instead of at the start of the main loop as in the client example or inside the server event handler as in the reverse proxy example).

**Comment by jbbcst** [comment]
ok thanks and I use the same `mg_mgr_poll(&mgr, 1000)`; and `mgr` for both client and server, as you told me in a previous answer?

**Reply by scaprile** [reply]
Unless you have a strong reason to have separate event managers, you are likely to have only one, as all examples do.

**Comment by jbbcst** [comment]
Hello,
Let's take this example of an  http-server:

```
int main(void) {
  struct mg_connection *c;

  mg_log_set(MG_LL_INFO);  // Set log level
  mg_mgr_init(&mgr);        // Initialize event manager
  mg_timer_add(&mgr, 15000, MG_TIMER_REPEAT | MG_TIMER_RUN_NOW, timer_fn,
               &mgr);                     // Init timer for demo purposes, 15s
  c = mg_listen(&mgr, s_lsn, sfn, NULL);  // Create server connection
  if (c == NULL) {
    MG_INFO(("SERVER cant' open a connection"));
    return 0;
  }
  while (true)
    mg_mgr_poll(&mgr, 100);  // Infinite event loop, blocks for upto 100ms
                             // unless there is network activity
  mg_mgr_free(&mgr);         // Free resources
  return 0;
}
```
but with ` struct mg_mgr mgr;  // Event manager` declared as global in main.c.
If I want to use the same mgr to create an http client in another part of the code (in another .c file so by using the global variable mgr), there might be a conflict?

**Comment by scaprile** [comment]
AFAIK there are no global variables in Mongoose if that is what you mean.
I would use a different name to ease debugging and preferably go for dependency injection instead of a global.

**Comment by jbbcst** [comment]
I agree with you that there are no global variables in Mongoose, I meant if I create an  http listener in the main function and I want to create an http client in another part of the code (in a c++ object which is independant), can I use the same mgr?

**Comment by scaprile** [comment]
As long as you keep everything in one thread you  shouldn't have issues, but can you post a small example of your idea ?

**Comment by jbbcst** [comment]
Yes of course, here is a small simplified example of the idea:

**main.cpp:**
```
struct mg_mgr mgr;  // Declared as global to share

int main(void) 
{
    // Create an object which be called by an event in the program
    auto httpClient = make_unique<HttpClient>();

    mg_mgr_init(&mgr);  // Initialise event manager    
    mg_http_listen(&mgr, s_listen_on, fn, NULL);  // Create HTTP listener
    for (;;) mg_mgr_poll(&mgr, 1000);             // Infinite event loop
    mg_mgr_free(&mgr);
    return 0;
}
```

**HttpClient.h:**

```
#pragma once
extern "C" {
#include "../../Mongoose/mongoose.h"
}
extern struct mg_mgr mgr;// Same mgr as in the main

class HttpClient
{
public:
    HttpClient();
    ~HttpClient();
private:
    void openHttpConnectionOnEvent(string dataToSend);
};
```

**httpClient.cpp:**
```
#include "HttpClient.h"
using namespace std;

HttpClient::HttpClient() {}

HttpClient::~HttpClient() {}

// This function is called by an event in the program
void HttpClient::openHttpConnectionOnEvent(string dataToSend)
{
    // When an event call this function, an http connect is created with the same mgr as in the main
    mg_http_connect(&mgr, "http://192.168.0.124:8000", handleHttpConnection, dataToSend) == NULL)
}

// Function which is called by mgr poll
void handleHttpConnection(struct mg_connection* c, int ev, void* ev_data, void* fn_data)
{
	string* dataToSend = (string*)fn_data;

	if (ev == MG_EV_OPEN) {
		// Connection created. Store connect expiration time in c->label
		*(uint64_t*)c->label = mg_millis() + 5000;
	}
	else if (ev == MG_EV_POLL) {
		if (mg_millis() > *(uint64_t*)c->label &&
			(c->is_connecting || c->is_resolving)) {
			mg_error(c, "Connect timeout");
		}
	}
	else if (ev == MG_EV_CONNECT) {
		// Connected to server. Extract host name from URL
		const char* uri = "http://192.168.0.124:8000";
		struct mg_str host = mg_url_host(uri);
		// If s_url is https://, tell client connection to use TLS
		if (mg_url_is_ssl(uri)) {
			struct mg_tls_opts opts = {};
			opts.ca = "ca.pem";
			opts.srvname = host;
			mg_tls_init(c, &opts);
		}
		// Send request
		char* postBody;
		postBody = mg_mprintf(
			R"({"data":%Q})",
			dataToSend);
		int content_length = postBody ? strlen(postBody) : 0;
		mg_printf(c,
			"%s %s HTTP/1.0\r\n"
			"Host: %.*s\r\n"
			"Content-Type: octet-stream\r\n"
			"Content-Length: %d\r\n"
			"\r\n",
			"POST", mg_url_uri(uri), (int)host.len,
			host.ptr, content_length);
		mg_send(c, postBody, content_length);
		free(postBody);
		postBody = nullptr;
	}
	else if (ev == MG_EV_HTTP_MSG) {
		// Tell mongoose to close this connection
		c->is_draining = 1;
	}
	else if (ev == MG_EV_ERROR) {
		c->is_closing = 1;
	}
}
```

**Comment by scaprile** [comment]
Can't you just pass mgr as a parameter to the constructor and hold it as an object variable ?
Who calls the function on an event ? You have an infinite loop and no one except another Mongoose event can call your function unless you use multithreading. Mongoose is not thread-safe, check the docs.

**Comment by jbbcst** [comment]
Yes I could pass mgr as parameter, the function `openHttpConnectionOnEvent` is called by a thread which communicate with the serial port.
I am  using multithreading to manage the listener but in this case above  where we want to create a client connection on event, I don't see clearly how to do it.

**Comment by scaprile** [comment]
From [the docs](https://mongoose.ws/documentation/#2-minute-integration-guide)
> NOTE: Since Mongoose's core is not protected against concurrent accesses, make sure that all mg_* API functions are called from the same thread or RTOS task.

If you absolutely need multi-threading, take the time to go through the user guide (linked above) and [tutorials](https://mongoose.ws/tutorials/), there is one that shows you how to do [multi-threading](https://mongoose.ws/tutorials/multi-threaded/).

**Comment by jbbcst** [comment]
Yes thank you, doc is well done, I understand and use the multi-threading from the event handler of mg_http_listen but in the case to create an mg_http_connect triggered by another thread , I don't see how to do it? Have you got a short example?
